internal/plugin/server.go

```diff
@@ -326,69 +326,70 @@ func (plugin *NvidiaDevicePlugin) Allocate(ctx context.Context, reqs *pluginapi.
 func (plugin *NvidiaDevicePlugin) getAllocateResponse(requestIds []string) (*pluginapi.ContainerAllocateResponse, error) {
 	deviceIDs := plugin.deviceIDsFromAnnotatedDeviceIDs(requestIds)
 
-	responseID := uuid.New().String()
-	response, err := plugin.getAllocateResponseForCDI(responseID, deviceIDs)
-	if err != nil {
-		return nil, fmt.Errorf("failed to get allocate response for CDI: %v", err)
+	// Create an empty response that will be updated as required below.
+	response := &pluginapi.ContainerAllocateResponse{
+		Envs: make(map[string]string),
 	}
 
+	if plugin.cdiEnabled {
+		responseID := uuid.New().String()
+		if err := plugin.updateResponseForCDI(response, responseID, deviceIDs...); err != nil {
+			return nil, fmt.Errorf("failed to get allocate response for CDI: %v", err)
+		}
+		return response, nil
+	}
+	if plugin.config.Sharing.SharingStrategy() == spec.SharingStrategyMPS {
+		plugin.updateResponseForMPS(response)
+	}
```

Comment:
Does the order of these need to be reversed? Otherwise we will never inject the MPS stuff when CDI is enabled, right?

Reply:
Ah ... sorry, the return shoudl not be in the cdiEnabled block.

Reply:
Hmm, now I'm more confused :)

When CDI is enabled we add the CDI devices AND still add the envars, mounts, etc. manually? That doesn't seem right...

I can buy it for the MPS envvars/mounts (since we don't have those in CDI), but I would assume we shouldn't rely on CDI as well as set them all ourselves.

Reply:
The issue is that we need to support workloads regardless of whether these use CDI or the NVIDIA container runtime -- put another way. In the case of the GPU operator where we have an `nvidia` and `nvidia-cdi` runtime class, the device plugin must inject the devices regardless of the mode. This is done by allowing *multiple* device-list-strategies to be enabled at the same time.

This means that if a device plugin has `cdi-annotations` set, it will set the contianer annotations. If -- in addition -- `envvars` is specified (or volume mounts for that matter), the envvars will be set.

If the workload is started with the CDI-enabled runtime, the annotations will be read and the `NVIDIA_VISIBLE_DEVICES` envvars ignored. If the workload is started by a non-CDI enabled runtime, the envvars are read as per normal and the annotations have no effect.

Reply:
OK, I think I can buy that. If the flags are set, we honor them. This makes sense for the various device list strategies as well as for `--pass-device-specs`. It's a little weird when applied to the `GDSEnabled` and `MOFEDEnabled` since we just say "enable me" but then go about setting the envvar both via CDI and manually (this would apply to the MPS flag in the future as well once we have CDI support for it).

In any case, I think it's fine, even if it seems a bit strange at first glance.

One thing I would change just to make the flow a bit more consistent is to move the following to the end amongst all of the other `config` checks:
```
	if plugin.config.Sharing.SharingStrategy() == spec.SharingStrategyMPS {
		plugin.updateResponseForMPS(response)
	}
``` 

and then change the check for `if plugin.cdiEnabled {` to `if plugin.deviceListStrategies.IsCDIEnabled() {` since this really is about checkin of the deviceListStrategy is asking for CDI or not (and it then matches the other calls made below).

Reply:
Just to clarify, we don't set those envvars for CDI, we actually inject a `nvidia.com/gds=all` device request (same for mofed).

The `plugin.cdiEnabled` change will require some additional refactoring in this case.

Reply:
I have made the change that you suggested and made some refactoring.

I have left it a separate commit so that we can move it to a separate PR if required.